qualtran/bloqs/chemistry/thc.ipynb

```diff
@@ -0,0 +1,302 @@
+    "Currently it is a good deal larger due to:\n",
+    " 1. inverting inequality tests should not need more toffolis.\n",
+    " 2. We are not using phase-gradient gate toffoli cost for Ry rotations\n",
+    " 3. Small differences in quoted vs implemented comparator costs.\n",
+    "\n",
+    "See: https://github.com/quantumlib/Qualtran/issues/390\n",
+    "\n",
+    "#### Parameters\n",
+    " - `num_mu`: THC auxiliary index dimension $M$\n",
+    " - `num_spin_orb`: number of spin orbitals $N$ \n",
+    "\n",
+    "#### Registers\n",
+    " - `mu`: $\\mu$ register.\n",
+    " - `nu`: $\\nu$ register.\n",
+    " - `succ`: ancilla flagging success of amplitude amplification.\n",
+    " - `eq_nu_mp1`: ancillas for flagging if $\\nu = M+1$. \n",
+    "\n",
+    "#### References\n",
+    "[Even more efficient quantum computations of chemistry through tensor hypercontraction](https://arxiv.org/pdf/2011.03494.pdf). Eq. 29.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1b6bb0e0",
+   "metadata": {
+    "cq.autogen": "_make_uniform_superposition.py"
+   },
+   "outputs": [],
+   "source": [
+    "from qualtran.bloqs.chemistry.thc import UniformSuperpositionTHC\n",
+    "\n",
+    "num_mu = 10\n",
+    "num_spin_orb = 4\n",
+    "bloq = UniformSuperpositionTHC(num_mu=num_mu, num_spin_orb=num_spin_orb)\n",
+    "show_bloq(bloq)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a89e358f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qualtran.resource_counting import get_bloq_counts_graph\n",
+    "from qualtran.drawing import GraphvizCounts\n",
+    "from qualtran.bloqs.chemistry.thc_notebook_utils import generalize\n",
+    "graph, sigma = get_bloq_counts_graph(bloq, generalizer=generalize)\n",
+    "GraphvizCounts(graph).get_svg()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f2cbc8e6",
+   "metadata": {},
+   "source": [
+    "Let's print out the costs contributing to the TCount."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fa6858a5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qualtran.bloqs.chemistry.thc_notebook_utils import bin_bloq_counts\n",
+    "\n",
+    "binned_counts = bin_bloq_counts(bloq)\n",
+    "# number of bits for mu register (nm in THC paper)\n",
+    "nm = bloq.signature[0].bitsize\n",
+    "# Costs for THC paper\n",
+    "# The factor of 4 is for Toffoli -> T conversion\n",
+    "paper_costs = {\n",
+    "    'comparator': sum([4*4*(nm - 1), 4*(4*nm - 3)]), # 4 comparitors of cost nm - 1 Toffolis\n",
+    "    'rotation': sum([4*4, 4*4]), # Given as br - 3, br = 7 is the number of bits of precision for rotations.\n",
+    "    'reflections': sum([4*3, 4*(2*nm-1)]), # 5 qubit reflection for comparitors and 2*nm + 1 qubits reflect after hadamards\n",
+    "    'other': sum([4*3]), # \"Checking the inequality test\" unclear if this is the multi-control not gate.\n",
```

Comment:
the `sum`s are weird here. I guess this is leftover from debugging? Just add the numbers together.

If you want to keep them factorized, then you'll need to re-work the datastructure to make it interpretable. I.e. make it clear which entry of the list is for which part. Or use a `set` if you don't care.

Reply:
I've removed the sums.